native/src/main/java/io/ballerina/lib/data/csvdata/utils/CsvUtils.java

```diff
@@ -0,0 +1,383 @@
+                    || tag == TypeTags.SIGNED32_INT_TAG || tag == TypeTags.UNSIGNED8_INT_TAG
+                    || tag == TypeTags.UNSIGNED16_INT_TAG || tag == TypeTags.UNSIGNED32_INT_TAG
+                    || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        if (csv instanceof BDecimal) {
+            if ((tag == TypeTags.DECIMAL_TAG
+                    || tag == TypeTags.FLOAT_TAG || tag == TypeTags.INT_TAG) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        if (csv instanceof Double) {
+            if ((tag == TypeTags.FLOAT_TAG
+                    || tag == TypeTags.DECIMAL_TAG || tag == TypeTags.INT_TAG) || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        if (Boolean.class.isInstance(csv)) {
+            if (tag == TypeTags.BOOLEAN_TAG || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        if (csv == null) {
+            if (tag == TypeTags.NULL_TAG || isJsonOrAnyDataOrAny(tag)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static boolean isJsonOrAnyDataOrAny(int tag) {
+        return tag == TypeTags.JSON_TAG || tag == TypeTags.ANYDATA_TAG || tag == TypeTags.ANY_TAG;
+    }
+
+    public static int getTheActualExpectedType(Type type) {
+        if (type instanceof TupleType) {
+            TupleType tupleType = (TupleType) type;
+            if (tupleType.getRestType() != null) {
+                return -1;
+            }
+            return tupleType.getTupleTypes().size();
+        } else {
+            return ((ArrayType) type).getSize();
+        }
+    }
+
+    public static Map<String, String> processNameAnnotationsAndBuildCustomFieldMap(RecordType recordType,
+                                                 Map<String, Field> fieldHierarchy) {
+        BMap<BString, Object> annotations = recordType.getAnnotations();
+        HashMap<String, String> updatedRecordFieldNames = new HashMap<>();
+        HashSet<String> updatedFields = new HashSet<>();
+        HashSet<String> updatedValues = new HashSet<>();
+
+        for (BString annotationsKey : annotations.getKeys()) {
+            String key = annotationsKey.getValue();
+            if (key.contains(Constants.FIELD)) {
+                BMap<BString, Object> annotMap = ((BMap<BString, Object>) annotations.get(annotationsKey));
+                for (BString mapKey : annotMap.getKeys()) {
+                    if (mapKey.getValue().endsWith(Constants.NAME)) {
+                        String name = ((Map<BString, Object>) annotMap.get(mapKey)).get(Constants.VALUE).toString();
+                        String originalName = key.substring(Constants.FIELD.length());
+                        if (updatedValues.contains(name)) {
+                            throw DiagnosticLog.error(DiagnosticErrorCode.DUPLICATE_FIELD, name);
+                        }
```

Comment:
I think you can combine both insert below and check here since if it cotained the value we anyway "return" from the method. This way you can avoid doing hashing and lookup twice for the same value. Same for `originalName` as well

```suggestion
                        if (!updatedValues.add(name)) {
                            throw DiagnosticLog.error(DiagnosticErrorCode.DUPLICATE_FIELD, name);
                        }
```

Reply:
I think the current way is more readable